osu.Game/Graphics/UserInterface/BarGraph.cs

```diff
@@ -5,76 +5,259 @@
+                    float length = maxLength == 0 ? 0 : Math.Max(0f, bar.Value / maxLength);
 
-                    if (bar.Bar != null)
+                    if (bar.Index < bars.Count)
                     {
-                        bar.Bar.Length = length;
-                        bar.Bar.Size = direction.HasFlagFast(BarDirection.Horizontal) ? new Vector2(1, size) : new Vector2(size, 1);
+                        bars.UpdateLength(bar.Index, length);
+                        continue;
                     }
-                    else
+
+                    bars.AddBar(length);
+                }
+
+                if (bars.Count > newCount)
+                    bars.RemoveRange(newCount, bars.Count - newCount);
+
+                animationStartTime = Clock.CurrentTime;
+                animationComplete = false;
+            }
+        }
+
+        private double animationStartTime;
+        private bool animationComplete;
+
+        private IShader shader = null!;
+        private Texture texture = null!;
+
+        [BackgroundDependencyLoader]
+        private void load(IRenderer renderer, ShaderManager shaders)
+        {
+            texture = renderer.WhitePixel;
+            shader = shaders.Load(VertexShaderDescriptor.TEXTURE_2, FragmentShaderDescriptor.TEXTURE);
+        }
+
+        protected override void Update()
+        {
+            base.Update();
+
+            if (!bars.Any)
+                return;
+
+            double currentTime = Clock.CurrentTime;
+
+            if (currentTime < animationStartTime + resize_duration)
+            {
+                bars.Animate(animationStartTime, currentTime);
+                Invalidate(Invalidation.DrawNode);
+            }
+            else if (!animationComplete)
+            {
+                bars.FinishAnimation();
+                Invalidate(Invalidation.DrawNode);
+
+                animationComplete = true;
+            }
+        }
+
+        protected override DrawNode CreateDrawNode() => new BarGraphDrawNode(this);
+
+        private class BarGraphDrawNode : DrawNode
+        {
+            public new BarGraph Source => (BarGraph)base.Source;
+
+            public BarGraphDrawNode(BarGraph source)
+                : base(source)
+            {
+            }
+
+            private IShader shader = null!;
+            private Texture texture = null!;
+            private Vector2 drawSize;
+            private BarDirection direction;
+            private float barBreadth;
+
+            private readonly List<float> lengths = new List<float>();
+
+            public override void ApplyState()
+            {
+                base.ApplyState();
+
+                shader = Source.shader;
+                texture = Source.texture;
+                drawSize = Source.DrawSize;
+                direction = Source.direction;
+                barBreadth = Source.bars.Breadth;
+
+                lengths.Clear();
+                lengths.AddRange(Source.bars.InstantaneousLengths);
```

Comment:
Not super sure this copying is required either. Could consider using a `Span` or something for zero-copy. Could be a decent win given that this'll be executing every frame?

Reply:
I think it's probably fine for now as is. Given the length can change based on new data this seems like the simplest implementation.